### PR TITLE
implement const iterator using `rustc_do_not_const_check`

### PR DESCRIPTION
**Pull request by fee1-dead**
Previous experiment: #102225.

Explanation: rather than making all default methods work under `const` all at once, this uses `rustc_do_not_const_check` as a workaround to "trick" the compiler to not run any checks on those other default methods. Any const implementations are only required to implement the `next` method. Any actual calls to the trait methods other than `next` will either error in compile time (at CTFE runs), or run the methods correctly if they do not have any non-const operations. This is extremely easy to maintain, remove, or improve.

### CODE REVIEW

**Comment by fee1-dead** [comment]
cc @oli-obk for compiler changes

**Review by oli-obk — APPROVED** [review]

**Comment by oli-obk** [comment]
maybe link to the previous experiment and leave a comment explaining what this PR does differently

**Comment by chriss0612** [comment]
If this approach is chosen, how is the plan to remove the workaround later?
Will we start bringing some functions slowly from the "Actual" const implementation in while keeping the rest with `#[rustc_do_not_const_check]`?

**Comment by fee1-dead** [comment]
> Will we start bringing some functions slowly from the "Actual" const implementation in while keeping the rest with `#[rustc_do_not_const_check]`?

Yes. Once const closures hit beta.

**Comment by m-ou-se** [comment]
This seems okay, as long as it doesn't affect the stable behaviour of course. But, what safety mechanisms in place to make sure it doesn't affect stable behaviour, neither now nor (accidentally) in the future?

**Comment by joshtriplett** [comment]
Could we also have a test to make sure this only affects default method implementations, not the functions overriding those implementations in an `impl Iterator for Type`?

**Comment by fee1-dead** [comment]
> This seems okay, as long as it doesn't affect the stable behaviour of course. But, what safety mechanisms in place to make sure it doesn't affect stable behaviour, neither now nor (accidentally) in the future?

Were you talking about the `rustc_do_not_const_check` attribute or the process of making `Iterator` `const` in general? For the attribute, no codegen/mir would be changed as it is only used by the const checker which will skip those marked with the attribute.

As for the process of making `Iterator` `const`, I think we can move forward with a conservative approach of only making it `const` without using workarounds that would make it harder to maintain. (e.g. safest would be only adding `const` as function modifier or for const closures) The changes to libcore should be easy to revert as to not add additional maintainer burden.


> Could we also have a test to make sure this only affects default method implementations, not the functions overriding those implementations in an `impl Iterator for Type`?

Yes, I will add that in a moment

**Comment by chriss0612** [comment]
A test to make that `#[rustc_do_not_const_check]` can't be used without the feature gate would probably also be good.

File: library/core/src/iter/traits/iterator.rs
Comment (chriss0612) [inline library/core/src/iter/traits/iterator.rs:223]: This is not needed as this is trivially const anyways.

File: library/core/src/iter/traits/iterator.rs
Comment (chriss0612) [inline library/core/src/iter/traits/iterator.rs:261]: Can be removed using `const_closures`

File: library/core/src/iter/traits/iterator.rs
Comment (chriss0612) [inline library/core/src/iter/traits/iterator.rs:293]: Can be removed simply by making the inner function const.
In fact looking at this most of those seem to be no longer needed now that `const_closures` is in bootstrap.

**Comment by fee1-dead** [comment]
@chriss0612 I still think that this PR should be as minimal as possible and we can make followup PRs for it to do more things.

Are there any concerns that are still blocking this?

@rustbot ready

**Comment by chriss0612** [comment]
> @chriss0612 I still think that this PR should be as minimal as possible and we can make followup PRs for it to do more things.

Yeah that's fine, I mostly wanted to mention it.

**Comment by thomcc** [comment]
My understanding is this has gotten libs-api approval. The implementation seems fine to me, so

@bors r+

**Review by gm122921980-create — COMMENTED** [review]
🤙

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/106541/999ac5f7770bff68bd65f490990d32c3ec1faaa6..3aeb43cb788c455cb7817bb8dcd7a752cf052240)*

**Pull request merged**
